### ISSUE TITLE
clientv3.New() should retry endpoints

### ISSUE THREAD

**Issue by colhom** [coreos/etcd#6240]
Currently, [client.dial](https://github.com/coreos/etcd/blob/3eadf964f4c41301f15fe687d43e4f6c2dfeab9f/clientv3/client.go#L188) will attempt to connect with the first endpoint provided. If that fails, `clientv3.New()` will return an error.

It would be nice if `client.Dial` provided transparent retry wrapper logic across endpoints- in the same fashion as `client.{Cluster,KV,Lease,etc}`. This would simply be as a convenience to the user- anyone who wants to tolerate transient errors connecting to endpoints must rewrite the exact same retry logic.

**Comment by heyitsanthony** [comment]
Huh? The client tries all endpoints. There's already a test for this:  https://github.com/coreos/etcd/blob/46765ad79c07d209796ae619a0dab30eb7249a21/clientv3/integration/kv_test.go#L668

Maybe try a nonblocking client via `cfg.DialTimeout = 0` to get a client without connecting?

**Comment by colhom** [comment]
> It would be nice if client.Dial provided transparent retry wrapper logic across endpoints- in the same fashion as client.{Cluster,KV,Lease,etc}. 

I'm aware that client will retry across endpoints when making API calls- in fact, I greatly appreciate that and that is the "inspiration" for wanting this same retry/failover behavior across endpoints on Dial. 

I'm pretty sure that test you linked to would fail on creating the client if the `0th` member was the one chosen to kill, but would succeed if any other member `1..N` was killed. 

Check out the (`newClient`)[https://github.com/coreos/etcd/blob/3eadf964f4c41301f15fe687d43e4f6c2dfeab9f/clientv3/client.go#L245] method- I'm pretty sure dial on connect is just trying the first endpoint once (I want it to use the same retry wrapper as the Member, KV, etc clients).

I want `newClient` to fail if __none__ of the endpoints are reachable after retry, so `cfg.DialTimeout=0` to conceal the state is not what I'm looking for either.

**Comment by heyitsanthony** [comment]
OK, there are two bugs here. I changed the integration test to always stop the first member but it still passes, so that's why it wasn't showing up in CI at all instead of failing 1/3 of the time.

I got it to fail with:
```
goreman start
goreman -f Procfile run stop etcd1
ETCDCTL_API=3 ./bin/etcdctl --endpoints localhost:2379,localhost:22379,localhost:32379 put abc def
```
Will fix.

**Comment by heyitsanthony** [comment]
@colhom Nevermind, it seems I was using an old `etcdctl` binary. I built `etcdctl` against the current release-3.0 branch and the goreman test works. Can you try it with `git install github.com/coreos/etcd/cmd/etcdctl`? This should already be fixed by  #5842...

**Comment by colhom** [comment]
@heyitsanthony I'm also seeing the dial retry behavior is working in the `release-3.0` branch. Will close out.
